test/sig_utils.py

```diff
@@ -301,3 +312,365 @@ def reference_vector_argument(arg):
+        self.__is_reverse_mode = any(arg.is_reverse_mode() for arg in args)
+
+    def is_reverse_mode(self):
+        return self.__is_reverse_mode
+
+    def is_matrix_like(self):
+        raise Exception("is_matrix_like not implemented for FunctionCallAssign")
+
+    def cpp(self):
+        return f"auto {self.name} = stan::math::eval({self.function_name}({self.arg_str}));"
+
+class FunctionCall(CppStatement):
+    def __init__(self, function_name, *args):
+        self.function_name = function_name
+        self.arg_str = ','.join(arg.name for arg in args)
+        self.__is_reverse_mode = any(arg.is_reverse_mode() for arg in args)
+
+    def is_reverse_mode(self):
+        return self.__is_reverse_mode
+
+    def cpp(self):
+        return f"{self.function_name}({self.arg_str});"
+
+class ExpressionArgument(CppStatement):
+    def __init__(self, overload, name, arg):
+        self.overload = overload
+        self.name = name
+        self.counter = IntArgument(f"{self.name}_counter", 0)
+        self.arg = arg
+    
+    def is_reverse_mode(self):
+        return self.arg.is_reverse_mode()
+
+    def is_matrix_like(self):
+        return True
+    
+    def is_expression(self):
+        return True
+
+    def cpp(self):
+        scalar = overload_scalar[self.overload]
+        counter_op_name = f"{self.name}_counter_op"
+
+        code = (
+            self.counter.cpp() + os.linesep +
+            f"stan::test::counterOp<{scalar}> {counter_op_name}(&{self.counter.name});" + os.linesep
+        )
+
+        if self.arg.stan_arg == "matrix":
+            return code + f"auto {self.name} = {self.arg.name}.block(0,0,1,1).unaryExpr({counter_op_name});"
+        elif self.arg.stan_arg in ("vector", "row_vector"):
+            return code + f"auto {self.name} = {self.arg.name}.segment(0,1).unaryExpr({counter_op_name});"
+        else:
+            raise Exception(f"Can't make an expression out of a {self.arg.stan_arg}")
+
+class FunctionGenerator:
+    def __init__(self, signature):
+        self.return_type, self.function_name, self.stan_args = parse_signature(signature)
+        self.reset()
+
+    def reset(self):
+        self.code_list = []
+    
+    def number_arguments(self):
+        return len(self.stan_args)
+
+    def is_ode(self):
+        return self.function_name.startswith("ode")
+    
+    def is_high_order(self):
+        return any("=>" in arg for arg in self.stan_args)
+
+    def is_rng(self):
+        return self.function_name.endswith("_rng")
+
+    def has_vector_arg(self):
+        return any(arg in eigen_types for arg in self.stan_args)
+
+    def returns_int(self):
+        return "int" in self.return_type
+
+    def add(self, statement):
+        if not isinstance(statement, CppStatement):
+            raise Exception("Argument to FunctionGenerator.add must be an instance of an object that inherits from CppStatement")
+
+        self.code_list.append(statement)
+        return statement
```

Comment:
Metzhods `add` and `cpp` seem totally unrelated to what the rest of this class is doing. So I think they should not be part of this class.

Reply:
Hmm, the idea of the class was to parse function signatures (to encapsulate a lot of the logic of the parsed functions in a Python class, so `fg.is_rng()` instead of doing `function_name`.endswith("_rng")) and then also manage the code generation.

I guess these could be two things. I'll think if that makes things clearer.